great_expectations/datasource/fluent/sql_datasource.py

```diff
@@ -712,6 +719,73 @@ def build_batch_request(
+        validate_partitioner: bool = True,
+    ) -> BatchDefinition[ColumnPartitioner]:
+        if validate_partitioner and partitioner:
+            self.validate_batch_definition(partitioner)
+        return super().add_batch_definition(name, partitioner)
+
+    @public_api
+    def validate_batch_definition(self, partitioner: ColumnPartitioner) -> None:
+        """Validates that the batch definition column is of a permissible type
+
+        This isn't meant to be called directly. This is called internally when a batch definition
+         is added. Data asset implementers can override this for their specific data asset.
+
+        Raises:
+            SqlAddBatchDefinitionError: The specified column to partition on is not of
+                a permissible type for batching (ie date or datetime) or no data is
+                present in this column.
+        """
+        # We only support certain partitioners for using as batch definitions.
+        assert isinstance(
+            partitioner,
+            (
+                ColumnPartitionerYearly,
+                ColumnPartitionerMonthly,
+                ColumnPartitionerDaily,
+            ),
+        )
+        # A _SQLAsset must have a SQLDatasource
+        assert isinstance(self.datasource, SQLDatasource)
+        engine: sqlalchemy.Engine = self.datasource.get_engine()
+
+        # It would be better to introspect the database types and see which ones map to date or
+        # datetime. However, 3rd party types, such as Snowflakes TIMESTAMP_NTZ haven't implemented
+        # all the sqlalchemy abstract methods such as `python_type`.
+        # To make this more concrete I would have liked to do something like:
+        # insp = sqlalchemy.inspect(self.datasource.get_engine())
+        # cols = insp.get_columns(self.table_name, self.schema_name)
+        # for col in cols:
+        #     pytype = col['type'].python_type
+        #
+        # Instead we query the db for a non-null value to see if sqlalchemy converts
+        # this value to a python date or datetime. This means we REQUIRE that data is
+        # present for this validation to work.
+        with engine.connect() as connection:
+            selectable: sqlalchemy.Selectable = self.as_selectable()
+            column: sqlalchemy.ColumnClause[Never] = sa.sql.column(partitioner.column_name)
+            try:
+                row = connection.execute(
+                    sa.select(column, selectable).limit(1)  # type: ignore[call-overload]  # sqlalchemy typing is missing variants
+                )
+            except Exception as query_error:
+                raise SqlAddBatchDefinitionError(
+                    msg=f"Attempt to read an example non-null '{column}' value from '{selectable}'"
+                    " failed so column type can't be verified to be a date or datetime."
+                ) from query_error
+
+            r = row.first()
+            if not r or not isinstance(r[0], (datetime, date)):
```

Comment:
Non-blocking, but `r[0]` works because we only select one column, right? Is there a way to make that more clear here, e.g. `r[partitioner.column_name]`? Not the end of the world, I just needed to think about it for an extra second.